frontend/src/metabase/querying/hooks/use-time-filter/use-time-filter.unit.spec.ts

```diff
@@ -0,0 +1,289 @@
+
+      const { result } = renderHook(() =>
+        useTimeFilter({
+          query,
+          stageIndex,
+          column,
+          filter,
+        }),
+      );
+
+      act(() => {
+        const { setValues } = result.current;
+        setValues(values);
+      });
+
+      act(() => {
+        const { operator, values, getFilterClause } = result.current;
+        const newFilter = checkNotNull(getFilterClause(operator, values));
+
+        expect(Lib.displayInfo(query, stageIndex, newFilter)).toMatchObject({
+          displayName,
+        });
+      });
+    },
+  );
+
+  it.each<CoerceFilterCase>([
+    {
+      operator: "between",
+      values: [TIME_3, TIME_1],
+      displayName: "Time is 10:20 AM – 8:15 PM",
+    },
+  ])(
+    "should allow to coerce a filter: $displayName",
+    ({ operator, values, displayName }) => {
+      const { result } = renderHook(() =>
+        useTimeFilter({
+          query: defaultQuery,
+          stageIndex,
+          column,
+        }),
+      );
+
+      act(() => {
+        const { getFilterClause } = result.current;
+        const newFilter = checkNotNull(getFilterClause(operator, values));
+
+        expect(
+          Lib.displayInfo(defaultQuery, stageIndex, newFilter),
+        ).toMatchObject({
+          displayName,
+        });
+      });
+    },
+  );
+
+  it.each<ValidateFilterCase>([
+    {
+      operator: "<",
+      values: [null],
+    },
+    {
+      operator: ">",
+      values: [null],
+    },
+    {
+      operator: "between",
+      values: [null, null],
+    },
+  ])(
+    'should validate values for "$operator" operator',
+    ({ operator, values }) => {
+      const { result } = renderHook(() =>
+        useTimeFilter({
+          query: defaultQuery,
+          stageIndex,
+          column,
+        }),
+      );
+
+      act(() => {
+        const { setOperator, setValues } = result.current;
+        setOperator(operator);
+        setValues(values);
+      });
+
+      act(() => {
+        const { operator, values, isValid, getFilterClause } = result.current;
+        expect(isValid).toBeFalsy();
+        expect(getFilterClause(operator, values)).toBeUndefined();
+      });
+    },
+  );
+
+  it("should preserve values when switching operators", () => {
+    const { result } = renderHook(() =>
+      useTimeFilter({
+        query: defaultQuery,
+        stageIndex,
+        column,
+      }),
+    );
+
+    act(() => {
+      const { setValues } = result.current;
+      setValues([TIME_1]);
+    });
+
+    act(() => {
+      const { values, getDefaultValues, setOperator, setValues } =
+        result.current;
+      const newOperator = ">";
+      setOperator(newOperator);
+      setValues(getDefaultValues(newOperator, values));
+    });
+
+    act(() => {
+      const { operator, values, getFilterClause } = result.current;
+      const newFilter = checkNotNull(getFilterClause(operator, values));
+
+      expect(
```

Comment:
afaik `expect` should go outside `act` as there is no any side effect in it

Reply:
Fixed